core/lib/src/response/responder.rs

```diff
@@ -217,34 +219,142 @@ impl<'r> Responder<'r> for String {
+                let (start, end) = match ranges[0] {
+                    ByteRangeSpec::FromTo(mut start, mut end) => {
+                        if end < start {
+                            return Ok(
+                                Response::build()
+                                    .status(Status::RangeNotSatisfiable)
+                                    .header(AcceptRanges(vec![RangeUnit::Bytes]))
+                                    .finalize()
+                            );
+                        }
+                        if start > size {
+                            start = size;
+                        }
+                        if end > size {
+                            end = size;
+                        }
+                        (start, end)
+                    },
+                    ByteRangeSpec::AllFrom(mut start) => {
+                        if start > size {
+                            start = size;
+                        }
+                        (start, size - start)
+                    },
+                    ByteRangeSpec::Last(len) => {
+                        // we could seek to SeekFrom::End(-len), but if we reach a value < 0, that is an error.
+                        // but the RFC reads:
+                        //      If the selected representation is shorter than the specified
+                        //      suffix-length, the entire representation is used.
+                        let start = cmp::max(size - len, 0);
+                        (start, size - start)
+                    }
+                };
+
+                body.seek(io::SeekFrom::Start(start))
+                    .expect("Attempted to seek to the start of the requested range, but failed.");
+
+                return Ok(
+                    Response::build()
+                        .status(Status::PartialContent)
+                        .header(AcceptRanges(vec![RangeUnit::Bytes]))
+                        .header(ContentRange(ContentRangeSpec::Bytes {
+                            range: Some((start, end)),
+                            instance_length: Some(end - start),
+                        }))
+                        .raw_body(Body::Sized(body, end - start))
+                        .finalize()
+                )
+            }
+            // A server MAY ignore the Range header field.
+        },
+        // An origin server MUST ignore a Range header field that contains a
+        // range unit it does not understand.
+        Range::Unregistered(_, _) => {},
+    };
+
+    Err(body)
+}
+
 /// Returns a response with Content-Type `application/octet-stream` and a
 /// fixed-size body containing the data in `self`. Always returns `Ok`.
 impl<'r> Responder<'r> for &'r [u8] {
-    fn respond_to(self, _: &Request) -> response::Result<'r> {
+    fn respond_to(self, req: &Request) -> response::Result<'r> {
+        let mut body = Cursor::new(self);
+        
+        match try_range_response(req, body) {
```

Comment:
Both of the `respond_to` methods are harder to read than without `Range` support, mostly because of that awkward function signature. I might play around with this and try to make things a bit nicer.